### ISSUE TITLE
support for truetype fonts?

### ISSUE THREAD

**Issue by ameridas** [cosmos72/twin#27]
It looks from my initial experiments that truetype fonts are not supported.  If that true?  Any idea how hard it would be to add support?  Thanks!

**Comment by cosmos72** [comment]
That's correct.

To add them, one would need to make a copy of server/hw/hw_X11.{c,h} (or server/hw/hw_gfx.{c,h} if you also want pseudo-graphics) to some new files (for example server/hw/hw_xft.{c,h}) and replace all font-related structs and calls, as (but not limited to): XCharStruct, XFontStruct, XDrawImageString, XListFontsWithInfo, XFreeFontInfo, XLoadQueryFont, XFreeFont
using instead the Xft* equivalents.

Unexpected difficulties may arise if there is no one-to-one mapping from X* font structures and functions to equivalent Xft* ones.
I'd recommend limiting to monospace fonts at least initially, because twin assumes the display is just a grid of character cells.

If you have experience with Xft* truetype font API, the task should be relatively straightforward (I do not have experience with it).

**Comment by ameridas** [comment]
OK, I have done some work and it's working well enough that the rest is mostly cleanup.  If you're open to considering my patches, I have some questions...

1) The Xft routines handle traditional bitmap fonts as well as truetype fonts, so we have some options:
a) modify hw_X11.c and hw_gfx.c to use the Xft routines, removing the calls to the old legacy-only-font routines
b) make it a compile-time option for hw_X11.c and hw_gfx.c to use Xft routines or the original implementation
c) modify hw_X11.c and hw_gfx.c to include both the Xft and the original font implementations "side by side", and auto-detect which to use based on the type of font that is loaded
c) fork hw_X11.c and hw_gfx.c, and the user chooses the Xft version (or not) via the --hw=blah command-line arg

Do you have a preference?

2) I currently have X11_MonospaceFontScore and X11_FontIsDualWidth disabled b/c some of the info isn't readily available.  It looks like there are two things being done - detecting monospace font, and calculating a score for the font if it is monospace.  I'll have to do some research on how to identify a monospace font (maybe fontconfig?)  If we need to keep the scoring, it would be helpful to me for a description of what the scoring algorithm is attempting to do -- it's not clear to me from reading the code.

**Comment by cosmos72** [comment]
Good :) thanks a lot! 

1: I'd say go for option c) fork hw_X11.c and/or hw_gfx.c.  
  The reason is simple: to also keep a bare-bone hw_X11.c with minimal dependencies.  
  Since the new hw_xft.c (feel free to choose another name) will *not* try to minimize dependencies, you can also consider leaving hw_X11.c untouched and only forking hw_gfx.c, which uses libXpm for a little graphics on window borders

2.1. X11_MonospaceFontScore gives a score to a candidate font based on three criteria:
   a. is monospace
   b. as a size and aspect ratio as similar as possible to what the user requested from the command line, as for example --hw=X,fontsize=10x20
   c. looks "good" in a terminal, by raising the score of certain font names - the ones listed in the source code are a subjective (i.e. my) preference

2.2. Feel free to ignore X11_FontIsDualWidth for the moment - it is work in progress. If you are curious, it should return true if the font is "almost" monospace, i.e. it only has two possible glyph bounding boxes: "narrow", as ASCII letters and numbers, and "wide", as Chinese ideograms, with "wide" glyphs exactly twice as wide as "narrow" glyphs. But the rest of twin does not yet fully support "wide" glyphs

**Comment by ameridas** [comment]
I'm currently looking into X11_ShowCursor.  Is it true that the only way to exercise the "soft cursor" code would be if an app sends a TW_DPY_SetCursorType message.  At least that's the only call to SetCursorType I see that looks relevant - but it is difficult to be certain that the SetCursorType in FlushCursor() wouldn't cause a soft cursor under some circumstance.

Actually it would be nice if the "hard cursor" was the less used option since I haven't yet discovered an xft replacement for XFillRectangle using xor function...

BTW, I think 'if (type & 0x10) {' in X11_ShowCursor is a bug since inside that block it does  'if ((type & 0x20)' and 'if ((type & 0x40)' which could never be true.  Based on the conditional for 'VGA hw-like cursor', I'm guessing the soft cursor condition should be at least type & 0xf0 and perhaps 0xfffff0 since 'v = (HWCOL(V) | ((type >> 16) & 0xff)) ^ ((type >> 8) & 0xff);' insinuates that 'type' contains 24 bits of information.

**Comment by cosmos72** [comment]
I admit that I don't remember much of  that stuff... it's very old code, and git blame confirms most of it was not touched in 11 years (!)

It looks correct to me:
"soft cursor" will change the foreground and/or background colors at cursor position.
* it is enabled by bit 4, i.e. (type & 0x10) != 0. In such case, the colors to use are the original cell colors with some bit flips (util.h:71)
* if bit 5 is **also** set, i.e. (type & 0x20) != 0 and the bit flipping above did not change the foreground color, then inverts the foreground color
* if bit 6 is **also** set, i.e. (type & 0x40) != 0 and the bit flipping above did not change the background color, then inverts the background color
* finally, sets the computed colors (util.h: 76...79) and draws the character at cursor position (util.h 80...83)
* note: it's important to store the currently set X11 colors into xsgc.foreground and xsgc.background , otherwise subsequent draws may get confused and use the wrong colors.

"VGA hw-like cursor" is simpler, and draws a small rectangle in GXxor mode with foreground color =  current foreground XOR current background.
* again, it's important to store the currently set X11 foreground color into xsgc.foreground

**Comment by cosmos72** [comment]
About a replacement for XFillRectangle using GXxor function:

I guess it could work as it is, provided that XDrawImageString16 is replaced with equivalent Xft font drawing function.   
If instead it doesn't work, we can use some other cursor look-and-feel, as for example a simple GXcopy

**Comment by ameridas** [comment]
re: that possible bug - you're right.  I had gotten myself confused.  Too late at night.

**Comment by ameridas** [comment]
If I fork hw_X11.c, there is going to be a ton of duplicate code between hw_X11.c and hw_xft.c.  My software engineering background is grating against that amount of duplicate code...  Would you be open to some code reorganization to pull out the common parts that are easily extracted?  Similar in concept to what you've already done with hw_x/util.h.  Of course I will fork and leave hw_X11.c alone if you really prefer it, but I thought it would be worth asking.  Thanks!

**Comment by cosmos72** [comment]
If you have all these energies to spare, sure :)

**Comment by ameridas** [comment]
OK, it's very close.  Mainly need to track down a couple of autoconf issues.  A few more questions...

I've never checked code into github.  It looks like the process is to make a branch that I check the code into, and then you review the branch and merge to master - is that correct?  I presume that you would need to make the branch and give me permission to check in...

Many of the truetype fonts I've used for testing are missing some of the special characters you use for window frame and background.  Could you give me a pointer to where those are listed and some advice on how I might do a runtime substitution for the ones that aren't available in the chosen font?

My tool versions are slightly different from what you used to autoconf/automake the current master branch:
autoconf (GNU Autoconf) 2.69
aclocal (GNU automake) 1.15
automake (GNU automake) 1.15

Should I go ahead and submit what I have created with these versions of the tools, or would you like to do the autoconf/automake stuff yourself so the tool versions match what you've been using?

And, finally, a quick glance at the code looks like you use 4-space indents with 8-position tab stops.  But it isn't consistent on using tabs vs just spaces.  My default editor setup is to use spaces exclusively instead of rely on the correct tabstop setting.  In other words, I'll match 4-space indents, but my indents will never use tabs.  Does that sound OK?

Thanks!

**Comment by cosmos72** [comment]
That's great!

The typical procedure is that you fork the project on github, so that you have your own repository for it, then commit & push the changes to your repository, and finally create a pull request on the original project - see https://help.github.com/articles/creating-a-pull-request-from-a-fork/

The autotools versions currently used are:
autoconf (GNU Autoconf) 2.69
aclocal (GNU automake) 1.15.1
automake (GNU automake) 1.15.1
so they are newer than yours. I will run autoconf/automake stuff myself.

And finally, 4-space indents without any tabs are ok, don't worry.

**Comment by cosmos72** [comment]
Any news? Even partial code would be useful :)

**Comment by ameridas** [comment]
Sorry - been busy the last couple of weeks.  Read through the autoconf documentation and now it's slightly clearer than mud. ;)  But I think I understand it well enough to finish up without too much more work.  I'm hoping I can spend a few hours on it this weekend.

**Comment by cosmos72** [comment]
Good :)
As I said, feel free to contirbute your code even if it's not 100% finished and polished.

**Comment by ameridas** [comment]
FYI, I'm synced up to head against your master branch and have committed my changes to my https://github.com/ameridas/twin fork.  I have a bit more cleanup - hopefully can do the pull request in the next couple of days.

**Comment by cosmos72** [comment]
That's good news!

**Comment by ameridas** [comment]
So...  When I went to fix my indentation, there were enough places where things weren't lining up that I took the liberty to just re-do the indentation on the three files where I'd made significant changes.  I apologize if it seems rude -- that was certainly not my intent.  If you prefer, I will revert those commits and fix my indentation by hand to line up as best I can with the existing indentation.

**Comment by cosmos72** [comment]
Fixing indentation is well known to create a lot of noise in diff's output.
Not to mention the amount of flame wars on the different styles...

Better if you leave as-is the indentation of existing code at the moment - diff's output will be much cleaner.

For the new code you wrote, feel free to indent it as you prefer: if needed, it will be "fixed" too by a (hopefully automated) future commit that enforces a consistent indentation style on all the code.

P.S. I had a quick look at your xft driver source code and tried using it. It worked fine, kudos :)

A minor note is the font autoselection algorithm - on my system, it chose a math font that had no ASCII characters, so I had to manually specify a font. I guess it could be improved, but it can be done after merging your driver - and consensus on what is a "good" font is at least partially subjective.

Waiting for the pull request :)

**Comment by ameridas** [comment]
Ya, I knew I shouldn't do that reformatting.  It's prohibited at my job for the reasons you mentioned as well as others.  It tends to be more of a problem an large projects, but it's not a big deal.  After reverting, I was able to match the indenting fairly easily once I took a few minutes to configure vim for the 4-space indents.

It's ready for you to take a look, so I'll do the pull request.  I split out much of the driver-specific code from the common code using functions with driver-specific implementations, but I did leave a few ifdefs where I felt it was reasonable.  Let me know if there are any left that you feel should be eliminated using the same technique.

Thanks for your patience, and I'm looking forward to hearing your feedback.

BTW, I didn't address the issue you mentioned about it choosing a default font that was all symbols.  I agree it would be better to handle that in a future iteration.  There are a few other things I have in mind to address in the future as well.
